gpt_index/indices/vector_store/index/base.py

```diff
@@ -0,0 +1,344 @@
+        self,
+        index_struct: IndexDict,
+        document: BaseDocument,
+        text_splitter: TokenTextSplitter,
+    ) -> None:
+        """Add document to index."""
+        nodes = self._get_nodes_from_document(document, text_splitter)
+        embedding_results = self._get_node_embedding_results(nodes, set(), document.get_doc_id())
+
+        new_ids = self._vector_store.add(embedding_results)
+
+        for result, new_id in zip(embedding_results, new_ids):
+            index_struct.add_node(result.node, text_id=new_id)
+
+    def _build_index_from_documents(self, documents: Sequence[BaseDocument]) -> IndexDict:
+        """Build index from documents."""
+        text_splitter = self._prompt_helper.get_text_splitter_given_prompt(
+            self.text_qa_template, 1
+        )
+        index_struct = self.index_struct_cls()
+        for d in documents:
+            self._add_document_to_index(index_struct, d, text_splitter)
+        return index_struct
+
+    def _insert(self, document: BaseDocument, **insert_kwargs: Any) -> None:
+        """Insert a document."""
+        self._add_document_to_index(self._index_struct, document, self._text_splitter)
+
+    def _delete(self, doc_id: str, **delete_kwargs: Any) -> None:
+        """Delete a document."""
+        self._index_struct.delete(doc_id)
+        self._vector_store.delete(doc_id)
+
+    @classmethod
+    def load_from_dict(cls, result_dict: str, **kwargs: Any) -> "BaseGPTIndex":
+        """Load index from string (in JSON-format).
+
+        This method loads the index from a JSON string. The index data
+        structure itself is preserved completely. If the index is defined over
+        subindices, those subindices will also be preserved (and subindices of
+        those subindices, etc.).
+
+        NOTE: load_from_string should not be used for indices composed on top
+        of other indices. Please define a `ComposableGraph` and use
+        `save_to_string` and `load_from_string` on that instead.
+
+        Args:
+            index_string (str): The index string (in JSON-format).
+
+        Returns:
+            BaseGPTIndex: The loaded index.
+
+        """
+        if 'vector_store' in result_dict:
+            config_dict = result_dict[VECTOR_STORE_CONFIG_DICT_KEY]
+        return super().load_from_dict(result_dict, **kwargs, **config_dict)
+
+    def save_to_dict(self, **save_kwargs: Any) -> dict:
+        """Save to string.
+
+        This method stores the index into a JSON string.
+
+        NOTE: save_to_string should not be used for indices composed on top
+        of other indices. Please define a `ComposableGraph` and use
+        `save_to_string` and `load_from_string` on that instead.
+
+        Returns:
+            dict: The JSON dict of the index.
+
+        """
+        out_dict = super().save_to_dict()
+        out_dict[VECTOR_STORE_CONFIG_DICT_KEY] = self._vector_store.config_dict
+        return out_dict
+
+    def query(
```

Comment:
Out of curiosity, why did you override this function? 

This will break being able to compose vector store in a graph

Reply:
This is because we now directly pass the vector store instance to `*Query` classes.

Currently `*Query` classes only has access to `IndexStruct` instances, which is quite limiting. This is a hack to get around that limitation without blowing up the scope of this refactor.

Reply:
you can pass vector store to query classes through _preprocess_query